server/lib/api/users.js

```diff
@@ -96,3 +139,71 @@ async function resetPassword(ctx, next) {
     ctx.status = 204
   })
 }
+
+async function verifyEmail(ctx, next, userSockets) {
+  const { id } = ctx.params
+  const { code } = ctx.query
+  const { email } = ctx.request.body
+
+  if (!id || !code || !isValidEmail(email)) {
+    throw new httpErrors.BadRequest('Invalid parameters')
+  }
+
+  const emailVerified = await useEmailVerificationCode(id, email, code)
+  if (!emailVerified) {
+    throw new httpErrors.BadRequest('Email verification code is invalid')
+  }
+
+  ctx.status = 204
+
+  // If the user is not logged in, our job is done here. They will receive the updated data with
+  // `emailVerified` set to `true` next time they log in.
+  if (!ctx.session.userId) return
+
+  // If the user is logged in when verifying their email, update all of their sessions to indicate
+  // that their email is now indeed verified.
+  const userSessionsKey = 'user_sessions:' + ctx.session.userId
+  const userSessionIds = await redis.smembers(userSessionsKey)
+
+  for (const sessionId of userSessionIds) {
+    // NOTE(2Pac): There is actually a race condition here (we first get the value, then modify and
+    // store it back; something else could modify it between the get and the store). However, fixing
+    // it would be quite involved (would probably need to modify the redis store) and the impact of
+    // it is not that great, so it's whatever.
+    const session = await sessionStore.get(sessionId)
+    if (session) await sessionStore.set(sessionId, { ...session, emailVerified: true })
+  }
+
+  // Above updates the sessions directly in the store, but the current session on the server remains
+  // unupdated. So we update it manually here.
+  ctx.session.emailVerified = true
+
+  // Last thing to do is to notify all of the user's opened sockets that their email is now verified
+  // NOTE(2Pac): With the way the things are currently set up on client (their socket is not
+  // connected when they open the email verification page), the client actually making the request
+  // won't get this event. Thankfully, that's easy to deal with on the client-side.
+  const user = userSockets.getByName(ctx.session.userName)
+  if (user) {
+    user.subscribe('/auth/emailVerification', () => ({ action: 'emailVerified' }))
```

Comment:
This is a really absurd way to do this, users should all be subscribed to a relevant path by default (one that has an identifying piece of information like their username or ID), and you should publish to that route when it changes. Don't abuse subscriptions as a way to publish

Reply:
Yeah, sorry about that; I tried to cheat a bit here :d

Reply:
I tried to subscribe a user to `/users/:userId` path for this, but then I saw we're already using that route for user's subscriptions: https://github.com/ShieldBattery/ShieldBattery/blob/master/client/loading/socket-handlers.js#L42

Should I reuse that path for notifying user's their email has been verified, or should I create a new path? I assume the latter, since we could probably reuse this path for all user's profile updates. Just not sure what would the appropriate path be :d